packages/core/src/agent/getOutboundMessageContext.ts

```diff
@@ -0,0 +1,247 @@
+import type { AgentMessage } from './AgentMessage'
+import type { AgentContext } from './context'
+import type { ConnectionRecord, Routing } from '../modules/connections'
+import type { BaseRecord } from '../storage/BaseRecord'
+
+import { Key } from '../crypto'
+import { ServiceDecorator } from '../decorators/service/ServiceDecorator'
+import { AriesFrameworkError } from '../error'
+import { OutOfBandService, OutOfBandRole, OutOfBandRepository } from '../modules/oob'
+import { OutOfBandRecordMetadataKeys } from '../modules/oob/repository/outOfBandRecordMetadataTypes'
+import { RoutingService } from '../modules/routing'
+import { DidCommMessageRepository, DidCommMessageRole } from '../storage'
+import { uuid } from '../utils/uuid'
+
+import { OutboundMessageContext } from './models'
+
+/**
+ * Maybe these methods should be moved to a service, but that would require
+ * extra injection in the sender functions, and I'm not 100% sure what would
+ * be the best place to host these.
+ */
+
+/**
+ * Get the outbound message context for a message. Will use the connection record if available,
+ * and otherwise try to create a connectionless message context.
+ */
+export async function getOutboundMessageContext(
+  agentContext: AgentContext,
+  {
+    message,
+    connectionRecord,
+    associatedRecord,
+    lastReceivedMessage,
+    lastSentMessage,
+  }: {
+    connectionRecord?: ConnectionRecord
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    associatedRecord?: BaseRecord<any, any, any>
+    message: AgentMessage
+    lastReceivedMessage?: AgentMessage
+    lastSentMessage?: AgentMessage
+  }
+) {
+  // TODO: even if using a connection record, we should check if there's an oob record associated and this
+  // is the first response to the oob invitation. If so, we should add the parentThreadId to the message
+  if (connectionRecord) {
+    agentContext.config.logger.debug(
+      `Creating outbound message context for message ${message.id} with connection ${connectionRecord.id}`
+    )
+    return new OutboundMessageContext(message, {
+      agentContext,
+      associatedRecord,
+      connection: connectionRecord,
+    })
+  }
+
+  if (!lastReceivedMessage) {
+    throw new AriesFrameworkError(
+      'No connection record and no lastReceivedMessage was supplied. For connection-less exchanges the lastReceivedMessage is required.'
+    )
+  }
+
+  if (!associatedRecord) {
+    throw new AriesFrameworkError(
+      'No associated record was supplied. This is required for connection-less exchanges to store the associated ~service decorator on the message.'
+    )
+  }
+
+  // Connectionless
+  return getConnectionlessOutboundMessageContext(agentContext, {
+    message,
+    associatedRecord,
+    lastReceivedMessage,
+    lastSentMessage,
+  })
+}
+
+export async function getConnectionlessOutboundMessageContext(
```

Comment:
Not for reusability but more for readability, it would be good to extract a lot of logic from this into separate functions. I am trying to understand the whole flow and got confused at certain parts because I did not know in which if statement / level I was reading. We could also switch to `tabWidth: 8` :).